File: Olympiad/COCI/official/2008/contest6_solutions/princeza.cpp
#include <algorithm>
#include <cstdio>

using namespace std;

#define MAXLOPOC 100000
#define MAXSKOK 100000

int n, k;

struct point {
   int x, y;
   int d1, d2;
   int skok[4];
} lopoci[MAXLOPOC];

int sorted[MAXLOPOC];
bool cmp_d1( int i, int j ) { return (lopoci[i].d1 != lopoci[j].d1) ? lopoci[i].d1 < lopoci[j].d1 : lopoci[i].d2 < lopoci[j].d2; }
bool cmp_d2( int i, int j ) { return (lopoci[i].d2 != lopoci[j].d2) ? lopoci[i].d2 < lopoci[j].d2 : lopoci[i].d1 < lopoci[j].d1; }

char niz[MAXSKOK+1];

int main( void ) {
   scanf( "%d%d ", &n, &k );
   gets( niz );
   for( int i = 0; i < n; ++i ) {
      scanf( "%d%d", &lopoci[i].x, &lopoci[i].y );
      lopoci[i].d1 = lopoci[i].x + lopoci[i].y;
      lopoci[i].d2 = lopoci[i].x - lopoci[i].y;

      for( int j = 0; j < 4; ++j ) lopoci[i].skok[j] = -1;

      sorted[i] = i;
   }
   sort( sorted, sorted + n, cmp_d1 );
   for( int i = 1; i < n; ++i ) {
      if( lopoci[sorted[i-1]].d1 == lopoci[sorted[i]].d1 ) {
         lopoci[sorted[i-1]].skok[1] = sorted[i];
         lopoci[sorted[i]].skok[2] = sorted[i-1];
      }
   }
   sort( sorted, sorted + n, cmp_d2 );
   for( int i = 1; i < n; ++i ) {
      if( lopoci[sorted[i-1]].d2 == lopoci[sorted[i]].d2 ) {
         lopoci[sorted[i-1]].skok[0] = sorted[i];
         lopoci[sorted[i]].skok[3] = sorted[i-1];
      }
   }

   int barica = 0;
   for( char *c = niz; *c; ++c ) {
      int next = lopoci[barica].skok[*c-'A'];
      if( next == -1 ) continue;

      for( int j = 0; j < 4; ++j )
         if( lopoci[barica].skok[j] != -1 )
            lopoci[lopoci[barica].skok[j]].skok[3-j] = lopoci[barica].skok[3-j];
      barica = next;
   }

   printf( "%d %d\n", lopoci[barica].x, lopoci[barica].y );

   return 0;
}

File: Olympiad/COCI/official/2008/contest6_solutions/princeza.pas
program princeza;

uses classes;

const MAXLOPOC = 100000;
const MAXSKOK = 100000;

type
   ppoint = ^point;
   point = record
      x, y : longint;
      d1, d2 : longint;
      skok : array[0..3] of ppoint;
   end;

function cmp_d1( A, B : pointer ) : longint;
var
   aa, bb : ppoint;
begin
   aa := A; bb := B;
   if aa^.d1 <> bb^.d1 then cmp_d1 := aa^.d1 - bb^.d1
                       else cmp_d1 := aa^.d2 - bb^.d2;
end;

function cmp_d2( A, B : pointer ) : longint;
var
   aa, bb : ppoint;
begin
   aa := A; bb := B;
   if aa^.d2 <> bb^.d2 then cmp_d2 := aa^.d2 - bb^.d2
                       else cmp_d2 := aa^.d1 - bb^.d1;
end;

var
   n, k : longint;
   i, j : longint;
   barica, next : ppoint;
   A, B : ppoint;

   lopoci : array[1..MAXLOPOC] of point;
   sorted_d1, sorted_d2 : TList;
   skokovi : array[1..MAXSKOK] of char;

begin
   readln( n, k );
   for i := 1 to k do read( skokovi[i] );
   readln;

   sorted_d1 := TList.Create;
   sorted_d2 := TList.Create;

   for i := 1 to n do begin
      readln( lopoci[i].x, lopoci[i].y );
      lopoci[i].d1 := lopoci[i].x + lopoci[i].y;
      lopoci[i].d2 := lopoci[i].x - lopoci[i].y;

      for j := 0 to 3 do lopoci[i].skok[j] := NIL;

      sorted_d1.add( @lopoci[i] );
      sorted_d2.add( @lopoci[i] );
   end;

   sorted_d1.Sort( @cmp_d1 );
   sorted_d2.Sort( @cmp_d2 );

   for i := 2 to n do begin
      A := sorted_d1[i-2];
      B := sorted_d1[i-1];
      if A^.d1 = B^.d1 then begin
         A^.skok[1] := B;
         B^.skok[2] := A;
      end;
   end;
   for i := 2 to n do begin
      A := sorted_d2[i-2];
      B := sorted_d2[i-1];
      if A^.d2 = B^.d2 then begin
         A^.skok[0] := B;
         B^.skok[3] := A;
      end;
   end;

   barica := @lopoci[1];
   for i := 1 to K do begin
      next := barica^.skok[ord(skokovi[i])-65];
      if next = NIL then continue;

      for j := 0 to 3 do
         if barica^.skok[j] <> NIL then
            barica^.skok[j]^.skok[3-j] := barica^.skok[3-j];
      barica := next;
   end;

   writeln( barica^.x, ' ', barica^.y );
end.
